contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,140 @@
+    '''
+    This plugin posts a GIF in response to the keywords provided by the user.
+    Images are provided by Giphy, through the public API.
+    The bot looks for messages starting with "@giphy" or @mention of the bot
+    and responds with a message with the GIF based on provided keywords.
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def triage_message(self, message, client):
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            return client.full_name != message['sender_full_name']
+
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        bot_response = get_bot_giphy_response(message, client)
+
+        if message['type'] == 'private':
+            client.send_message(dict(
+                type='private',
+                to=message['sender_email'],
+                content=bot_response,
+            ))
+        else:
+            client.send_message(dict(
+                type='stream',
+                to=message['display_recipient'],
+                subject=message['subject'],
+                content=bot_response,
+            ))
+
+
+class GiphyNoResultException(Exception):
+    pass
+
+
+def get_giphy_api_key_from_config():
+    config = SafeConfigParser()
+    with open(os.environ['HOME'] + '/.giphy_config', 'r') as config_file:
+        if six.PY2:
+            config.readfp(config_file)
+        else:
+            config.read_file(config_file)
+    return config.get("giphy-config", "key")
+
+
+def get_url_gif_giphy(keyword, api_key):
+    # Return a URL for a Giphy GIF based on keywords given.
+    # In case of error, e.g. failure to fetch a GIF URL, it will
+    # return a number.
+    query = {'s': keyword,
+             'api_key': api_key}
+    try:
+        data = requests.get(GIPHY_TRANSLATE_API, params=query)
+    except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+        logging.warning(e)
+        raise
+
+    search_status = data.json()['meta']['status']
+    if search_status != 200 or data.status_code != 200:
+        raise requests.exceptions.ConnectionError
+
+    try:
+        gif_url = data.json()['data']['images']['original']['url']
+    except (TypeError, KeyError):  # Usually triggered by no result in Giphy.
+        raise GiphyNoResultException
```

Comment:
like `raise GiphyNoResultException()` so that it creates an instance of that class.